### PR TITLE
Give a per user smarts log file

### PR DESCRIPTION
**Pull request by Gamenot**
This to address #157

### CODE REVIEW

File: smarts/core/scenario.py [hunk 1]
```diff
@@ -715,6 +715,16 @@ def bubbles(self):
     def mission(self, agent_id):
         return self._missions.get(agent_id, None)
 
+    def _resolve_log_dir(self, log_dir):
+        home_dir = os.path.expanduser("~")
+        _, home_user = os.path.split(home_dir)
+        if log_dir is None:
+            log_dir = os.path.abspath(f"/tmp/smarts_{home_user}/_sumo_run_logs")
```
Comment (davidrusu) [inline smarts/core/scenario.py:722]: `/tmp` is not something we can rely on to be cross platform.

Use `mkdtemp` with a `SMARTS_` prefix instead. https://docs.python.org/3.9/library/tempfile.html#tempfile.mkdtemp

**Review by davidrusu — CHANGES_REQUESTED** [review]
/tmp is not cross platform

File: smarts/core/scenario.py (hunk 1 above)
Comment (liamchzh) [reply smarts/core/scenario.py:722]: +1

File: smarts/core/utils/file.py [hunk 2]
```diff
@@ -82,3 +85,26 @@ def file_md5_hash(file_path: str) -> str:
...
+
+    # if not os.path.exists(temp_dir):
+    #     original_mask = os.umask(0)
+    #     try:
+    #         os.makedirs(temp_dir, mode=0o777)
+    #         os.chmod(temp_dir, mode=0o777)
+    #     finally:
+    #         os.umask(original_mask)
+
+    # return os.path.join(temp_dir, f"{home_user}")
```
Comment (Gamenot) [inline smarts/core/utils/file.py:105]: This above would work on linux and macos. The question would be windows.

File: smarts/core/scenario.py [hunk 3]
```diff
@@ -22,9 +22,10 @@
 import logging
 import math
 import os
-import sys
 import pickle
 import random
+import stat
+import tempfile
```
Comment (liamchzh) [inline smarts/core/scenario.py:28]: Are these 2 imports being used?

File: smarts/sstudio/generators.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
 import os
 import random
 import logging
+from smarts.core.utils.file import make_dir_in_smarts_log_dir
```
Comment (liamchzh) [inline smarts/sstudio/generators.py:23]: This line should be in the 3rd part of imports.

File: smarts/core/utils/file.py [hunk 5]
```diff
@@ -17,10 +17,13 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
+from functools import lru_cache
 import hashlib
 import os
 import shutil
 import dataclasses
+import stat
```
Comment (liamchzh) [inline smarts/core/utils/file.py:25]: I believe `stat` is not used.

File: smarts/sstudio/generators.py (hunk 4 above)
Comment (Gamenot) [reply smarts/sstudio/generators.py:23]: Sorry, I have automatic imports that tend to throw the imports in random places. :facepalm:

File: smarts/core/utils/file.py (hunk 2 above)
Comment (liamchzh) [reply smarts/core/utils/file.py:104]: Can `tempfile` take care of the permission?

File: smarts/sstudio/generators.py (hunk 4 above)
Comment (liamchzh) [reply smarts/sstudio/generators.py]: Seems you have a new tool to help re-order the imports. I like that re-ordering.

File: smarts/core/scenario.py [hunk 6]
```diff
@@ -22,27 +22,32 @@
...
 from functools import lru_cache
-from itertools import cycle, product
+from itertools import cycle
+from itertools import product
 from pathlib import Path
-from typing import Any, Dict, Sequence, Tuple
+from typing import Any
+from typing import Dict
+from typing import Sequence
+from typing import Tuple
```
Comment (davidrusu) [inline smarts/core/scenario.py:37]: why are we spreading out these imports across multiple lines?

File: smarts/core/utils/file.py (hunk 2 above)
Comment (davidrusu) [reply smarts/core/utils/file.py:104]: If the intention is to have a one directory that is consistent across runs, then we shouldn't rely on /tmp/

Let's create a `.smarts` directory in the user's home directory and put sumo logs under there.

File: smarts/core/scenario.py (hunk 6 above)
Comment (Gamenot) [reply smarts/core/scenario.py:37]: I used a formatter and the justification for it is that it allows for easier navigation of merges.

Comment (davidrusu) [reply smarts/core/scenario.py:37]: Wait, you used a formatter other than black! Heathen!!

In all seriousness, I do prefer the old format of grouping all modules from the same parent on one line.

File: smarts/core/utils/file.py (hunk 2 above)
Comment (Gamenot) [reply smarts/core/utils/file.py:104]: > Can `tempfile` take care of the permission?
> 
> Sorry, I am not sure what's the problem here. I mean if `tempfile.mkdtemp` works well, why would we need to handle the permission?

There only consideration is that we are changing the file structure from the original to be completely flat. If that is already fine then there are no other issues.

Comment (Gamenot) [reply smarts/core/utils/file.py:105]: > If the intention is to have a one directory that is consistent across runs, then we shouldn't rely on /tmp/
> 
> Let's create a `.smarts` directory in the user's home directory and put sumo logs under there.

That can be done. The commented out solution uses a combination of the home directory name and `tempfile.gettempdir()`(which finds the same temp directory used by `tempfile.mkdtemp`.) The original reason for using temp over the home directory was to allow for automatic cleanup of the logs, which we could suspect can become quite large.

File: smarts/core/scenario.py (hunk 6 above)
Comment (Gamenot) [reply smarts/core/scenario.py]: Fair enough.

File: smarts/core/utils/file.py [hunk 7]
```diff
@@ -82,3 +85,15 @@ def file_md5_hash(file_path: str) -> str:
         hasher.update(f.read().encode())
 
     return str(hasher.hexdigest())
+
+
+@lru_cache(maxsize=1)
+def smarts_log_dir() -> str:
+    ## Following should work for linux and macos
+    smarts_dir = os.path.join(os.path.expanduser("~"), "smarts")
```
Comment (liamchzh) [inline smarts/core/utils/file.py:93]: `.smarts`?

**Review by JVillella — APPROVED** [review]
Assuming we use local dir over temp as the conversation was going, rest looks good to me

File: smarts/core/utils/file.py [hunk 8]
```diff
@@ -82,3 +85,15 @@ def file_md5_hash(file_path: str) -> str:
         hasher.update(f.read().encode())
 
     return str(hasher.hexdigest())
+
+
+@lru_cache(maxsize=1)
```
Comment (davidrusu) [inline smarts/core/utils/file.py:90]: No need for a cache here.

**Review by davidrusu — APPROVED** [review]
Just get rid of the lru cache, besides that LGTM

**Pull request merged**